### ISSUE TITLE
wslgit show -- inconsistent truncated output

### ISSUE THREAD

**Issue by diablodale** [andy-5/wslgit#28]
`wslgit show` is producing output that is truncated to an unpredictable length. I discovered this by using WSLgit in VS Code and noticing the visual diffs were inconsistent.

## Setup

Windows 10 build 17074.1002 ⬅ this insider build may be the cause of this issue
WSLgit v0.50 (precompiled exe from github)

## Repro

1. Install and configure WSLgit as per the instructions in this same repo.
2. Ensure WSLgit is in your path or be prepared to give the full pathname.
3. Clone the WSLgit repo to your local Windows machine
4. Change directory into the repro's top directory
5. Run `wslgit show HEAD:src/main.rs` repeatedly and verify the output

## Result

Eventually you will see the output is not the complete file. It will be truncated.
You may need to run step 5 many times (20+) until you see this problem.
I request someone try this on the general release of Win10 FCU build 16299.248

## Expected

Always the complete file.

## Notes

I put in some debug code and recompiled WSLgit. I can see that the `vec[u8] output.stdout` directly after the `wait_with_output()` will have different lengths. When it is truncated, the lengths are always a power of 2 like these: 512, 64, 1024, 256, etc.

I do not experience any truncation with the following:
* Run within WSL `git --no-pager show HEAD:src/main.rs`
* Run within CMD `wsl git --no-pager show HEAD:src/main.rs`

This suggests to me that there is a bug that is timing-, buffer-, or memory-related somewhere in the handoff of WSL's git output from WSL->Win32->Rust->WSLgit. It is unknown if the bug is in Windows, in Rust's compiler, or in the WSLgit codebase. However, my instinct tells me the problem is in Win32->Rust->WSLgit. Why? Because there is no truncation when the Win32 app called CMD.exe displays the data. That leaves Rust and WSLgit as the remaining components.

**Comment by andy-5** [comment]
Thanks for the detailed report.

I tried it on Win10 FCU build 16299.248 and could not reproduce this behaviour (I tried the main.rs source file, as suggested, and a long ~1000 lines text file).

Does the commit https://github.com/diablodale/wslgit/commit/977c8a9b2e45c6698a7091ba111699fe5456916c in your fork resolve this issue? From that commit it looks like the reason is non UTF-8 characters in the file, is that correct? Or is it just that stdout is not flushed?

However, I'm not sure the unsafe casting of the stdout bytes into a Rust string is the way to go. It seems this could cause problems, because for the path translation the string is assumed to be a valid UTF-8 Rust string, which might not be true in that case.

**Comment by diablodale** [comment]
Perhaps that's a good sign you can't replicate the problem. If you could not reproduce the problem by trying 20+ times, then it is probably a bug in the Windows Insider build. Microsoft hasn't sent an updated build to my slow ring in almost two months. I just now switched to the fast ring (from slow) and a new build is downloading.

No, none of the code I've committed to github fix this issue. I grinded several hours yesterday on this issue since it is hitting me every time I use the Code editor. I did find the cause of the power-of-2 size and the code that is truncating; both are in core Rust libraries. I copied the library code to my own private function and could trace behavior. The bug is occurring when Win32 hands the stream of stdout over to Rust's io handlers. Stdout is being truncated or not buffered enough for Rust to read the full contents before the process dies. Rust is reading... rust expands its buffer by power of 2 to read more... and on the next read there is no further data so the Rust library passes the incomplete data on to the Rust app.

I *can* workaround this issue by making a very large receive buffer in Rust, e.g. 32k. The buffer has to be large enough to contain the full byte size of what stdout is sending. This is not really a great solution as the problem will occur yet again when someone shows/diff's a very large file, e.g. 33k.

I did find some articles in MSDN and the general internet which suggest that Win32 does full buffering of named-pipe stdout. I saw some workarounds to get overlapped IO and/or line buffering, but they all were very low-level APIs and something core Rust should manage.

And sorry for putting that utf-8 commit into the same branch. I should keep that in a separate branch associated with a separate utf-8 issue that I've written yet haven't posted yet. I'll see if I can retract that commit and push the retraction to github. Then open a new issue on utf-8 handling with a separate branch as I have some technical details already researched on it.

**Comment by diablodale** [comment]
:-(  Unfortunately Win10 insider build 17110 experiences the same truncation problem with both your official v0.50 compiled and my private compile.

I have two ideas to explore. I'll try them today.

**Comment by diablodale** [comment]
Good news.  Two things changed. After these, I can no longer reproduce this issue.

1. New approach for commands like `show`. Don't grab stdio in a pipe. Instead the child inherits it from the parent. See commit https://github.com/diablodale/wslgit/commit/1c0c442e99b7ded30727655c2ac7cf5fdad35c90
2. Windows Insider build 17115.1 was downloaded and installed.

I ran a test 500 times in a `show` loop checking for the exact content on each loop. In 500 there were zero failures. 🙂 Before these two changes, I would have a truncation in ten or less loops.

Caution: while this fixes it for `show`, the truncation issue may be hiding with other commands that use the Rust stdio codepath. So I tested two other commands. A loop of 500 passed for both the following; there were no truncations.

* `wslgit ls-tree -r -t -l --full-tree HEAD`
* `wslgit diff main.rs`

I recommend closing this issue for now.
